### PR TITLE
Implement Eve

### PR DESCRIPTION
**Pull request by muupan**
This PR will add a new optimizer Eve #1846 .

Since Eve requires loss values to compute step sizes, `GradientMethod.update` is overwritten to get the outputs of `lossfun`. It will raise an error if `lossfun` is not specified. Tests for optimizers are also modified to use `lossfun`.

I found it is better than Adam when applied to `examples/mnist/train_mnist.py` with default parameters:
Eve
```
epoch       main/loss   validation/main/loss  main/accuracy  validation/main/accuracy
1           0.193573    0.0952835             0.9419         0.9696
2           0.0725242   0.0800335             0.977466       0.9765
3           0.0345596   0.06424               0.988915       0.981
4           0.013741    0.0567913             0.995832       0.9843
5           0.00571857  0.0562762             0.998649       0.985
6           0.00299029  0.053972              0.999333       0.9859
7           0.00175658  0.0581067             0.99965        0.9857
8           0.00100978  0.0592461             0.999917       0.9856
9           0.000664501  0.0626751             0.999933       0.9861
10          0.000589156  0.0727927             0.999933       0.9853
11          0.0013305   0.0659082             0.999733       0.9859
12          0.00049436  0.0670607             0.999883       0.9865
13          0.000336688  0.0687204             0.99995        0.9858
14          0.00010929  0.0690714             1              0.986
15          6.87041e-05  0.0699605             1              0.9865
16          5.16802e-05  0.0734176             1              0.986
17          3.75787e-05  0.0766912             1              0.9856
18          0.0028124   0.0766279             0.999183       0.985
19          0.000150661  0.0755002             1              0.9855
20          7.40332e-05  0.076406              1              0.9857
```
Adam
```
epoch       main/loss   validation/main/loss  main/accuracy  validation/main/accuracy
1           0.192746    0.118016              0.9411         0.9631
2           0.0727157   0.093093              0.977249       0.97
3           0.0501551   0.0860783             0.983615       0.9738
4           0.0367548   0.0736718             0.988399       0.9783
5           0.0272907   0.0791892             0.990932       0.9786
6           0.0258209   0.0827545             0.991516       0.9796
7           0.0203185   0.0644044             0.993332       0.9828
8           0.0177427   0.0757816             0.994432       0.9822
9           0.0161435   0.0873263             0.994815       0.9803
10          0.0139228   0.0853089             0.995249       0.9811
11          0.01513     0.081602              0.995066       0.983
12          0.0108932   0.0992509             0.996665       0.9789
13          0.0152876   0.0997881             0.995049       0.9811
14          0.0105591   0.0815175             0.996832       0.9841
15          0.00797358  0.0912213             0.997833       0.983
16          0.0113328   0.102879              0.996566       0.9801
17          0.0126291   0.087415              0.996166       0.9836
18          0.00641422  0.0877847             0.998116       0.9839
19          0.00915058  0.117586              0.997432       0.9811
20          0.00941298  0.126669              0.997215       0.9775
```

### CODE REVIEW

File: chainer/optimizers/eve.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    """Eve optimization algorithm.
+
+    See: https://arxiv.org/abs/1611.01505
+
+    You must speficify ``lossfun`` when you call its ``update`` method so that
+    it can utilize loss values in optimization.
+    """
+
+    def __init__(self, alpha=0.001, beta1=0.9, beta2=0.999, beta3=0.999,
+                 eps=1e-8, k=0.1, K=10):
```
Comment (gwtnb) [inline chainer/optimizers/eve.py]: It is difficult to distinguish `k` from `K` at a glance.

Comment (muupan) [inline chainer/optimizers/eve.py]: That's true, but the original paper uses `k` and `K`. What name would be better? The authors are using `thl` and `thu` in their real code. Should we use them or more intuitive names such as `lower_threshold`?

Comment (gwtnb) [inline chainer/optimizers/eve.py]: I think `lower_threshold` is a good name. If you would like to make the code consistent with the paper, you can use `upper_k` and `lower_k`.

Comment (muupan) [inline chainer/optimizers/eve.py]: Renamed `k` to `lower_threshold`, `K` to `upper_threshold`.

**Comment by muupan** [comment]
I noticed that computation of `f` was wrong, so fixed by 774e4342a23d472dfeebc1a00185c5d3939167fa .

 Results after the fix:
```
epoch       main/loss   validation/main/loss  main/accuracy  validation/main/accuracy
1           0.189799    0.0971503             0.941683       0.9682
2           0.0659906   0.0682113             0.979349       0.978
3           0.0347293   0.066329              0.988398       0.9805
4           0.0163772   0.0560146             0.994799       0.9831
5           0.00829566  0.0649794             0.997616       0.9828
6           0.00445461  0.0608943             0.998816       0.9838
7           0.0026657   0.0627363             0.999316       0.9855
8           0.00287208  0.0694796             0.99915        0.984
9           0.0036473   0.077297              0.998866       0.9827
10          0.00386606  0.0791341             0.998883       0.9827
11          0.00195498  0.07248               0.9995         0.9845
12          0.00122861  0.0743349             0.9997         0.9845
13          0.000216642  0.0795301             0.999967       0.9848
14          0.0050487   0.1048                0.998366       0.9782
15          0.00225744  0.0767912             0.999333       0.984
16          0.00142096  0.0724954             0.999717       0.9858
17          0.000387077  0.0751545             0.999933       0.9857
18          8.67336e-05  0.0739607             1              0.9865
19          0.000771507  0.134043              0.999833       0.9757
20          0.00676599  0.0719798             0.997783       0.9865
```

**Comment by jayanthkoushik** [comment]
Hi, recently [an issue](https://www.reddit.com/r/MachineLearning/comments/5bsmvn/research_improving_stochastic_gradient_descent/) was brought up about the algorithm not being invariant to shifting. My coauthor and I believe it has a simple fix, which we are working on. This does not affect any of the results in the paper, and should not be a problem for common loss functions like MSE or cross-entropy; but it certainly needs to be addressed.

I am not aware of any other pull requests for Eve at the moment, so I thought I'd bring this up here. I will be sure to post an update once the algorithm has been suitably modified.

**Comment by bkvogel** [comment]
Hi jayanthkoushik,
Thank you for letting us know about the issue. Please let us know when you have an updated algorithm and we will then proceed with this pull request review.
